File: Data Structures & Algorithms/design-hashmap/submission-0.py
class MyHashMap:

    def __init__(self):
        self.LOAD_FACTOR = 0.75
        self.hash_table = [None] * 16
        self.capacity = 16
        self.size = 0
    
    def calc_hash(self, element: int) -> int:
        return element % self.capacity

    def put(self, key: int, value: int) -> None:
        def resize():
            new_capacity = self.capacity * self.capacity
            new_hash_table = [None] * new_capacity
            prev_hash_table = self.hash_table
            self.hash_table = new_hash_table
            self.capacity = new_capacity
            for e in prev_hash_table:
                if e is None:
                    continue
                self.add(e)

        if self.get(key) == value:
            print("25 " + self.get(key)) 
            return
        idx = self.calc_hash(key)
        while self.hash_table[idx] is not None and self.hash_table[idx][0] != key:
            idx = (idx + 1) % self.capacity
        self.hash_table[idx] = [key, value]
        self.size += 1
        if self.size >= self.capacity * self.LOAD_FACTOR:
            resize()

    def get(self, key: int) -> int:
        if self.size == 0:
            return -1
        idx = self.calc_hash(key)
        while self.hash_table[idx] is not None:
            if self.hash_table[idx][0] == key:
                return self.hash_table[idx][1]
            idx = (idx + 1) % self.capacity
        return -1

    def remove(self, key: int) -> None:
        if self.get(key) == -1:
            return
        idx = self.calc_hash(key)
        while self.hash_table[idx][0] != key:
            idx = (idx + 1) % self.capacity
        self.hash_table[idx] = None
        self.size -= 1
        idx = (idx + 1) % self.capacity
        while self.hash_table[idx]:
            element = self.hash_table[idx]
            self.hash_table[idx] = None
            self.size -= 1
            self.put(element[0], element[1])
            idx = (idx + 1) % self.capacity


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)

File: Data Structures & Algorithms/lfu-cache/submission-0.java
class LFUCache {
    private final int capacity;
    private final Map<Integer, Node> nodeMap;
    private final Map<Integer, DoublyLinkedList> listMap;
    private int size;
    private int lfuCount;

    public LFUCache(int capacity) {
        this.capacity = capacity;
        this.nodeMap = new HashMap<>(capacity);
        this.listMap = new HashMap<>(capacity);
        this.size = 0;
        this.lfuCount = 0;
    }

    public int get(int key) {
        if (!nodeMap.containsKey(key)) {
            return -1;
        }

        Node node = nodeMap.get(key);
        updateFrequency(node);
        return node.value;
    }

    public void put(int key, int value) {
        if (nodeMap.containsKey(key)) {
            Node node = nodeMap.get(key);
            node.value = value;
            updateFrequency(node);
        } else {
            if (this.size == this.capacity) {
                DoublyLinkedList list = listMap.get(this.lfuCount);
                Node node = list.removeFromHead();
                this.nodeMap.remove(node.key);
                --this.size;
            }
            Node node = new Node(key, value);
            nodeMap.put(key, node);
            listMap.computeIfAbsent(1, k -> new DoublyLinkedList()).addToTail(node);
            this.lfuCount = 1;
            this.size++;
        }
    }

    private void updateFrequency(Node node) {
        DoublyLinkedList list = listMap.get(node.freq);
        list.remove(node);
        if (node.freq == lfuCount && list.size == 0) {
            lfuCount++;
        }
        int newFreq = node.incrementFreq();
        this.listMap.computeIfAbsent(newFreq, k -> new DoublyLinkedList()).addToTail(node);
    }

    private class Node {
        int key, value, freq;
        Node left, right;

        {
            this.freq = 1;
        }

        Node() {}

        Node(int key, int value) {
            this.key = key;
            this.value = value;
        }

        int incrementFreq() {
            this.freq++;
            return this.freq;
        }
    }

    private class DoublyLinkedList {
        private Node head;
        private Node tail;
        private int size;

        DoublyLinkedList() {
            this.head = new Node();
            this.tail = new Node();
            this.head.right = this.tail;
            this.tail.left = this.head;
        }

        void addToTail(Node node) {
            node.left = this.tail.left;
            this.tail.left.right = node;
            node.right = this.tail;
            this.tail.left = node;
            this.size++;
        }

        void remove(Node node) {
            node.left.right = node.right;
            node.right.left = node.left;
            node.left = null;
            node.right = null;
            this.size--;
        }

        Node removeFromHead() {
            Node node = this.head.right;
            this.remove(node);
            return node;
        }
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */